src/annotation/decorator_utils.ts

```diff
@@ -1,67 +1,99 @@
+        if(duplicate !== undefined) {
+            throw new Error(`${ERROR_MSGS.DUPLICATED_METADATA} ${duplicate.toString()}`);
+        }
+    }else{
+        metadatas = [metadata];
+    }
+    return metadatas;
 }
 
 function _tagParameterOrProperty(
     metadataKey: string,
     annotationTarget: any,
-    propertyName: string,
-    metadata: interfaces.Metadata,
-    parameterIndex?: number
+    key: string | symbol,
+    metadata: interfaces.MetadataOrMetadataArray,
 ) {
+    const metadatas: interfaces.Metadata[] = _ensureNoMetadataKeyDuplicates(metadata);
 
-    let paramsOrPropertiesMetadata: interfaces.ReflectResult = {};
-    const isParameterDecorator = (typeof parameterIndex === "number");
-    const key: string = (parameterIndex !== undefined && isParameterDecorator) ? parameterIndex.toString() : propertyName;
-
-    // if the decorator is used as a parameter decorator, the property name must be provided
-    if (isParameterDecorator && propertyName !== undefined) {
-        throw new Error(ERROR_MSGS.INVALID_DECORATOR_OPERATION);
-    }
-
+    let paramsOrPropertiesMetadata:any = {};
     // read metadata if available
     if (Reflect.hasOwnMetadata(metadataKey, annotationTarget)) {
         paramsOrPropertiesMetadata = Reflect.getMetadata(metadataKey, annotationTarget);
     }
 
-    // get metadata for the decorated parameter by its index
-    let paramOrPropertyMetadata: interfaces.Metadata[] = paramsOrPropertiesMetadata[key];
+    let paramOrPropertyMetadata: interfaces.Metadata[] | undefined = paramsOrPropertiesMetadata[key];
 
-    if (!Array.isArray(paramOrPropertyMetadata)) {
+    if (paramOrPropertyMetadata === undefined) {
         paramOrPropertyMetadata = [];
     } else {
         for (const m of paramOrPropertyMetadata) {
-            if (m.key === metadata.key) {
+            if (metadatas.some(md => md.key === m.key)) {
                 throw new Error(`${ERROR_MSGS.DUPLICATED_METADATA} ${m.key.toString()}`);
             }
         }
     }
 
     // set metadata
-    paramOrPropertyMetadata.push(metadata);
+    paramOrPropertyMetadata.push(...metadatas);
     paramsOrPropertiesMetadata[key] = paramOrPropertyMetadata;
     Reflect.defineMetadata(metadataKey, paramsOrPropertiesMetadata, annotationTarget);
 
 }
+type ParameterOrPropertyDecorator = (
+    target: any,
+    targetKey: string | symbol,
+    indexOrPropertyDescriptor?: number | TypedPropertyDescriptor<unknown>
+) => void;
+function createTaggedDecoratorInternal(
+    metadata:interfaces.MetadataOrMetadataArray,
+    callback?:ParameterOrPropertyDecorator
+) {
+    const decorator:ParameterOrPropertyDecorator = (target, targetKey, indexOrPropertyDescriptor) => {
+        if(callback){
+            callback(target, targetKey, indexOrPropertyDescriptor);
+        }
+        if (typeof indexOrPropertyDescriptor === "number") {
+            tagParameter(target, targetKey as unknown as string | symbol | undefined, indexOrPropertyDescriptor, metadata);
```

Comment:
I think this cast is not necessary

Reply:
It is as target key cannot be undefined for decorator

Reply:
done